modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java

```diff
@@ -122,8 +122,34 @@
         Netty4Utils.setup();
     }
 
+    private static final ByteSizeValue MTU_ETHERNET = new ByteSizeValue(1500);
+
+    private static final String SETTING_KEY_HTTP_NETTY_MAX_COMPOSITE_BUFFER_COMPONENTS = "http.netty.max_composite_buffer_components";
+
     public static Setting<Integer> SETTING_HTTP_NETTY_MAX_COMPOSITE_BUFFER_COMPONENTS =
-        Setting.intSetting("http.netty.max_composite_buffer_components", -1, Property.NodeScope);
+        new Setting<>(SETTING_KEY_HTTP_NETTY_MAX_COMPOSITE_BUFFER_COMPONENTS, (s) -> {
+            ByteSizeValue maxContentLength = SETTING_HTTP_MAX_CONTENT_LENGTH.get(s);
+            // Netty accumulates buffers containing data from all incoming network packets that make up one HTTP request in an instance of
+            // io.netty.buffer.CompositeByteBuf (think of it as a buffer of buffers). Once its capacity is reached, the buffer will iterate
+            // over its individual entries and put them into larger buffers (see io.netty.buffer.CompositeByteBuf#consolidateIfNeeded()
+            // for implementation details). We want to to resize that buffer because this leads to additional garbage on the heap and also
+            // increases the application's native memory footprint (as direct byte buffers hold their contents off-heap).
+            //
+            // With this setting we control the CompositeByteBuf's capacity (which is by default 1024, see
+            // io.netty.handler.codec.MessageAggregator#DEFAULT_MAX_COMPOSITEBUFFER_COMPONENTS). To determine a proper default capacity for
+            // that buffer, we need to consider that the upper bound for the size of HTTP requests is determined by `maxContentLength`. The
+            // number of buffers that are needed depend on how often Netty reads network packets which depends on the network type (MTU).
+            // We assume here that Elasticsearch receives HTTP requests via an Ethernet connection which has a MTU of 1500 bytes.
+            //
+            // Note that we are *not* pre-allocating any memory based on this setting but rather determine the CompositeByteBuf's capacity.
+            // The tradeoff is between less (but larger) buffers that are contained in the CompositeByteBuf and more (but smaller) buffers.
+            // With the default max content length of 100MB and a MTU of 1500 bytes we would allow 69905 entries.
+            long maxBufferComponentsEstimate = Math.round((double) (maxContentLength.getBytes() / MTU_ETHERNET.getBytes()));
```

Comment:
We can get the interface MTU from `NetworkInterface` API; should we use this here?

Reply:
I had a look this and I really like your idea. However, I am not sure it is feasible: I guess we need to resolve the network interface that is associated with the publish address (leveraging `NetworkService`). 

While I also dislike having the MTU specified as a constant here (instead of retrieving it from the network interface), it serves as an input for an estimation. The "worst" practical MTU is 1500 byte, another typical one is 65536 byte for loopback (that's not always the case for loopback but it is usually higher than for Ethernet). What happens if we assume an MTU of 1500 for our estimation in the loopback case? Our estimate would allow more buffer components than are expected for loopback. But in practice we do not expect to reach that theoretical capacity. A specific example for the loopback case (with a max content length of 100 MB):

* Estimated maximum number of buffers: 100 MB / 1500 byte = 69905
* Actual maximum number of buffers: 100 MB / 65536 byte = 1600

This means that we expect to reserve space for at most 1600 buffers in the `HttpObjectAggregator` but overestimate it to at most 69905 (which we do not reach in practice).

To summarize, I see two possibilities:

* Determine the publish address with `NetworkService`, find the matching network interface and determine its MTU. This is correct but more complex.
* Rename the constant `MTU_ETHERNET` to something like `SMALLEST_EXPECTED_MTU` (while [RFC 791](https://tools.ietf.org/html/rfc791) states 68 bytes as the minimum, I'd stick to 1500 for practical purposes). It's not ideal but in very special cases, users can explicitly define this expert setting (`http.netty.max_composite_buffer_components`) directly.

Reply:
@jasontedor wdyt about the above?

Reply:
I think that jumbo frames are common enough that we should try to go the extra mile here. If it's not possible to do cleanly, I would at least like to see a system property that can be set to set the MTU to 9000.

Reply:
Thanks for the suggestion. I have now introduced a system property with a default is 1500 (bytes).